### ISSUE TITLE
Different behavior between diff_iou_rotated_2d and box_iou_rotated

### ISSUE THREAD

**Issue by GlassyWing** [open-mmlab/mmcv#2335]
I'm confusing aboult the reuslts between two way.
```
pred_oc = torch.tensor([[ 0.0000,  0.0000,  5.0000, 10.0000,  np.pi / 6.]]
target_oc = torch.tensor([[ 0.0000,  0.0000, 10.0000,  5.0000,  np.pi / 3.]]

rois = diff_iou_rotated_2d(pred_oc[None], target_oc[None]) # Got `0`
rois = box_iou_rotated(pred_oc, target_oc, aligned=True) # Got `0.4058`
```

At least, the first result `0` mustn't be correct. Do they use a different angle representation？

**Comment by zytx121** [comment]
Could you try to use this PR to replace diff_iou_rotated_2d https://github.com/open-mmlab/mmcv/pull/2067

**Comment by GlassyWing** [comment]
> According to our experience, ` diff_iou_rotated_2d` is very unstable. [lilanxiao/Rotated_IoU#39](https://github.com/lilanxiao/Rotated_IoU/issues/39)
> 
> Could you try to use this PR [https://github.com/open-mmlab/mmcv/pull/2067，](https://github.com/open-mmlab/mmcv/pull/2067%EF%BC%8C) and looking forward to your feedback.

I tried this PR, seems it works good. Now they behave the same，except when the coordinates have negative values:

```
pred_oc = torch.tensor([[ 0.0000,  0.0000,  -5.0000, 10.0000,  np.pi / 6.]]
target_oc = torch.tensor([[ 0.0000,  0.0000, -10.0000,  5.0000,  np.pi / 3.]]

rois = diff_iou_rotated_2d(pred_oc[None], target_oc[None]) # Got -0.1261`
rois = box_iou_rotated(pred_oc, target_oc, aligned=True) # Got `0`
```

I don't think that's a problem, but the second value looks more in line with iou's range of values.

**Comment by GlassyWing** [comment]
There are still some bugs in this funcation with this [pull ](https://github.com/open-mmlab/mmcv/pull/2067%EF%BC%8C)  It looks like there are problems with the gradient calculation
```
pred = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5).requires_grad_()
target = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5)

rois = diff_iou_rotated_2d(pred, target)
loss = 1 - rois
loss.backward()
print(pred.grad) # tensor([[[ 0.0000e+00,  0.0000e+00, -2.0000e-01, -1.0000e-01,  7.4506e-09]]],
       device='cuda:0')
```

Even if the two boxes overlap, there is still a gradient.

**Comment by zytx121** [comment]
> > According to our experience, ` diff_iou_rotated_2d` is very unstable. [lilanxiao/Rotated_IoU#39](https://github.com/lilanxiao/Rotated_IoU/issues/39)
> > Could you try to use this PR [https://github.com/open-mmlab/mmcv/pull/2067，](https://github.com/open-mmlab/mmcv/pull/2067%EF%BC%8C) and looking forward to your feedback.
> 
> I tried this PR, seems it works good. Now they behave the same，except when the coordinates have negative values:
> 
> ```
> pred_oc = torch.tensor([[ 0.0000,  0.0000,  -5.0000, 10.0000,  np.pi / 6.]]
> target_oc = torch.tensor([[ 0.0000,  0.0000, -10.0000,  5.0000,  np.pi / 3.]]
> 
> rois = diff_iou_rotated_2d(pred_oc[None], target_oc[None]) # Got -0.1261`
> rois = box_iou_rotated(pred_oc, target_oc, aligned=True) # Got `0`
> ```
> 
> I don't think that's a problem, but the second value looks more in line with iou's range of values.

We require the format of the rbox to be (x, y, h, w, angle), shape (n, 5).  It is meaningless when `w` or `h` is less than 0.

**Comment by zytx121** [comment]
> There are still some bugs in this funcation with this [pull ](https://github.com/open-mmlab/mmcv/pull/2067%EF%BC%8C) It looks like there are problems with the gradient calculation
> 
> ```
> pred = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5).requires_grad_()
> target = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5)
> 
> rois = diff_iou_rotated_2d(pred, target)
> loss = 1 - rois
> loss.backward()
> print(pred.grad) # tensor([[[ 0.0000e+00,  0.0000e+00, -2.0000e-01, -1.0000e-01,  7.4506e-09]]],
>        device='cuda:0')
> ```
> 
> Even if the two boxes overlap, there is still a gradient.

@filaPro Please have a look.

**Comment by GlassyWing** [comment]
> > There are still some bugs in this funcation with this [pull ](https://github.com/open-mmlab/mmcv/pull/2067) It looks like there are problems with the gradient calculation
> > ```
> > pred = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5).requires_grad_()
> > target = torch.tensor([10, 10, 5, 10, 0.]).cuda().reshape(1, 1, 5)
> > 
> > rois = diff_iou_rotated_2d(pred, target)
> > loss = 1 - rois
> > loss.backward()
> > print(pred.grad) # tensor([[[ 0.0000e+00,  0.0000e+00, -2.0000e-01, -1.0000e-01,  7.4506e-09]]],
> >        device='cuda:0')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Even if the two boxes overlap, there is still a gradient.
> 
> @filaPro Please have a look.

I already know what the problem is,  in function `box_intersection` where the “Convention: if two edges are collinear, there is no intersection point.” is not correctly, at this point, the intersection is the vertex.

Just change these two lines：

from

```
mask_t = (t >0) & (t <1)
...
mask_u = (u > 0) & (u < 1)
```
to
```
mask_t = (t >= 0) & (t <= 1)
...
mask_u = (u >= 0) & (u <= 1)
```

This gives the correct result.
